Adding testing documentation (#45)
cheerfulstoic/ecto_watch: guides/introduction/Testing.md

If you need to verify that your application correctly receives `ecto_watch` events, consider the following:  

- **PostgreSQL, not Ecto, emits the notifications.** This means that for your test to capture notifications, the transaction that triggers them **must be committed**.

- **By default, transactions in tests are not committed.** The [`Ecto.Adapters.SQL.Sandbox`](https://hexdocs.pm/ecto_sql/Ecto.Adapters.SQL.Sandbox.html) ensures test transactions are rolled back unless explicitly overridden.  

- **To ensure notifications are sent, disable the sandbox mode for the test:** 
```elixir
Ecto.Adapters.SQL.Sandbox.checkout(MyRepo, sandbox: false)

```
This forces all database changes to be persisted, so notifications will be emitted. However, be mindful that you may need to manually clean up test records.

### Example

```elixir

> defmodule My.Application do
>   use Application
> 
>   @impl true
>   def start(_type, _args) do
>     children = [
>      {EctoWatch,
>        repo: MyRepo,
>        pub_sub: MyPubSub,
>        watchers: [
>          {record, :updated,
>           trigger_columns: [:example_column], label: :notification}
>        ]}
>     ]

Reviewer: ```suggestion
defmodule MyApp.Application do
  use Application

  alias MyApp.Records

  @impl true
  def start(_type, _args) do
    children = [
     {EctoWatch,
       repo: MyRepo,
       pub_sub: MyPubSub,
       watchers: [
         {Records.Record, :updated}
       ]}
    ]
```

I took out `trigger_columns: [:example_column]` since it isn't used in the rest of the example.

I think we should take out the `label: :notification` as well and have the `handle_info` expect `{Records.Record, :updated}` because that's the base use case for `ecto_watch` that people start with.

    opts = [strategy: :one_for_one, name: My.Supervisor]
    Supervisor.start_link(children, opts)
  end
end
```

```elixir
# my_notifier.ex -> module to test
defmodule MyNotifier do
  use GenServer
  
  # Rest of the GenServer Code

  # Handle incoming notifications and update state
  def handle_info({:notification, %{id: id}}, state) do
    new_state = Map.update(state, :counter, 1, &(&1 + 1))
    {:noreply, new_state}
  end

  def handle_call(:get_state, _from, state), do: {:reply, state, state}
end

> ```

Reviewer: I like making examples realistic whenever I can.  It's a bit weird that this GenServer is a "notifier" but the only logic that we see is updating a counter.  So maybe the module could be called `UpdateCounter`?

Also, the `state` is a map, but we only see `counter` being used.  Perhaps the `state` could just be an integer for the sake of the example.  Then there could be a:

```elixir
  def handle_info({:notification, %{id: id}}, count) do
     {:noreply, count + 1}
   end

  def handle_call(:get_count, _from, count), do: {:reply, count, count}
```

Or if we wanted to make the example a bit more potentially realistic we could use the `id` and have the `GenServer` count the number of times each record has been updated:

```elixir
  def handle_info({:notification, %{id: id}}, counts) do
    {:noreply, Map.update(counts, id, 1, &(&1 + 1))}
  end

  def handle_call({:get_count, id, id}, _from, counts), do: {:reply, Map.get(counts, id), counts}

  def handle_call(:get_counts, _from, counts) do
    total =
      counts
      |> Map.values()
      |> Enum.sum()

    {:reply, total, counts}
  end
```

The `get_counts` isn't really neccesary for the example but it's fun 😅

Author: I like being fun! haha so I also include that to be test

in our tests

```elixir
# my_notifier_test.exs

> defmodule MyNotifierTest do
>   use ExUnit.Case
> 

Reviewer: ```suggestion
defmodule MyApp.MyNotifierTest do
  use ExUnit.Case

  alias MyApp.Records
```

  setup do
    # Ensure database changes are committed during tests
    Ecto.Adapters.SQL.Sandbox.checkout(Repo, sandbox: false)

    # Clean up database before running each test
    cleanup()
  end

  defp cleanup do
    # Add logic to remove test records from the database
  end

>   test "receives a notification when the image description is updated" do
>     # Subscribe to notifications for assertions
>     :ok = EctoWatch.subscribe(:notification)
> 
>     # Based on our application config the following should emit the notification
>     # event if committed
>     record = record_fixture(%{})
>     Records.update_record(record, %{key: "some_value"})
> 
>     assert_receive {:notification, %{id: ^record.id}}, 2000
> 
>     state = Notifier.get_state()
>     assert Map.get(state, :counter) == 1  # Ensure core logic were executed
>   end

Reviewer: ```suggestion
  test "counter increments whenever a record is updated" do
    assert Notifier.get_state().counter == 0

    # Based on our application config the following should emit the notification
    # event if committed
    record = record_fixture(%{})
    Records.update_record(record, %{key: "some_value"})

    # Ensure core logic was executed
    assert Notifier.get_state().counter == 1
  end
```

I think it's better to just test the `Notifier` GenServer because maybe the implementation of `Notifier` will change and not use `ecto_watch` 🤷

Moved the comment because it's what `mix format` would do.

Reviewer: Also, you reference the `Notifier` module here, but maybe you mean `MyNotifier`?
